Coding Style Cleanup

Mottie opened the pull request:
Sorry for this all-encompassing PR...

* As mentioned in #109.
  * Updated linting rules.
  * Added eslint check (added a `package.json`). Use `npm run lint` to check.
  * Fix *almost* all linting issues... I left some unused variable warnings alone.
    * Indentation with 2 spaces.
    * Single quotes.
    * Strict `===` and `!==` followed.
* Added an issue template.
* Added a contributing readme file.
* Created `edit.css` file from the stylesheet within `edit.html`
* Restructured folders layout. See #25.
  * If you zip the `src` folder, that will now act as the uploadable source for the web stores.
  * The new folder structure is as follows (not in alphabetical order):

    ```sh
    stylus
    ├── .github/   # contributing & issue_template files
    ├── tools/
    │    ├── fill_locale_placeholders.py
    │    ├── pull_locales.rb
    │    └── pull_locales.sh
    ├── src/
    │    ├── _locales/
    │    ├── edit/
    │    │    ├── edit.css
    │    │    └── edit.js
    │    ├── images/
    │    ├── manage/
    │    │    ├── manage.css
    │    │    └── manage.js
    │    ├── options/
    │    │    ├── index.css
    │    │    └── index.js
    │    ├── popup/
    │    │    ├── popup.css
    │    │    └── popup.js
    │    ├── vendor/
    │    │    ├── backup/
    │    │    ├── beautify/
    │    │    ├── codemirror/
    │    │    ├── codemirror-overwrites/
    │    │    ├── csslint/
    │    │    └── msgbox/
    │    ├── apply.js
    │    ├── background.js
    │    ├── dom.js
    │    ├── edit.html
    │    ├── index.html
    │    ├── install.js
    │    ├── localization.js
    │    ├── manage.html
    │    ├── manifest.json
    │    ├── messaging.js
    │    ├── popup.html
    │    ├── prefs.js
    │    ├── storage.js
    │    └── update.js
    └── {dot-files, package.json, etc}
    ```

tophf commented on .editorconfig:
```diff
@@ -4,6 +4,7 @@ root = true
 indent_style = space
 indent_size = 2
 tab_width = 2
+end_of_line = lf
```
We're using git autocrlf so this is not needed.

Mottie replied:
I'm using Windows, and oddly when I open some files in Atom, they have CRLF line endings. I have it set up to always save with unix line endings... this addition shouldn't cause any issues.

tophf replied:
The problem is that currently files have CRLF so changing this rule without converting all files will lead to mixed line endings.

tophf commented on .eslintrc:
```diff
@@ -58,6 +58,10 @@ globals:
   prefs: false
   setupLivePrefs: false
+  # edit.js
+  exports: false
```
No need to expose stuff from edit.js as it's local to editor only. Remove lines 61-64

tophf commented on .eslintrc:
```diff
@@ -249,11 +253,11 @@ rules:
   sort-keys: [0]
   space-before-blocks: [2, always]
-  space-before-function-paren: [2, never]
+  space-before-function-paren: [0, never]
```
Why? I hate it.

Mottie replied:
hehe ok... I'm not a mind reader 😉... I'll take care of it later today. I have to run now! 😸

tophf replied:
@schomery likes it though. Let's set it to `1`.

tophf commented on .eslintrc:
```diff
@@ -262,4 +266,4 @@ rules:
   wrap-iife: [2, inside]
   yield-star-spacing: [2, {before: true, after: false}]
-  yoda: [2, never]
+  yoda: [0]
```
Please! It's one of the most retarded things.

Mottie replied:
It's in the code... I didn't feel like modifying any code, just the linting... I'll work on it too!

tophf replied:
What's in the code?

Mottie replied:
In the `edit.js` file on lines 66, 67 (`"handled" == CodeMirror` ) and 83, 109 ( `"checkbox" == node.type`) - [ref](https://github.com/openstyles/stylus/blob/master/edit.js#L66-L109).

tophf replied:
Disregard edit.js "style" as it's nonexistent, 99% of this file is inherited from Stylish. There are 15 js files that more or less foloow the modern coding style.

tophf commented on src/background.js:
```diff
@@ -31,9 +31,11 @@ chrome.tabs.onAttached.addListener((tabId, data) => {
     if (tab.url.startsWith(URLS.ownOrigin + 'edit.html')) {
-      chrome.windows.get(tab.windowId, {populate: true}, win => {
+      chrome.windows.get(tab.windowId, {
+        populate: true
```
This robotic expansion of literals doesn't make the code more readable. On the contrary, it's less readable in such cases as this one where only one parameter is used.

tophf replied:
Well, I can overwrite the IDE setting but I'd rather not do it manually.

Mottie replied:
Hehe, I might have used "Atom beautify" a couple of times, so yeah it was literally a robotic expansion 😛. I'll return these to their previous state.

tophf replied:
Yes please. With the exception of edit.js all other files follow eslint rules (well, a few eslint errors surfaced after its recent update).

tophf commented on src/background.js:
```diff
@@ -49,20 +51,24 @@ if ('commands' in chrome) {
 {
-  const onInstall = ({reason}) => {
+  const onInstall = ({
+    reason
```
This robotic expansion of literals doesn't make the code more readable. On the contrary, it's less readable in such cases as this one where only one parameter is used.

tophf commented on src/background.js:
```diff
@@ -49,20 +51,24 @@ if ('commands' in chrome) {
     // reset L10N cache on UI language change or update
-    const {browserUIlanguage} = tryJSONparse(localStorage.L10N) || {};
+    const {
+      browserUIlanguage
```
This robotic expansion of literals doesn't make the code more readable. On the contrary, it's less readable in such cases as this one where only one parameter is used.

tophf commented on src/background.js:
```diff
@@ -88,14 +94,18 @@ if ('commands' in chrome) {
   chrome.runtime.onInstalled.addListener(onInstall);
-  setTimeout(onInstall, 60e3, {reason: 'unbindme'});
+  setTimeout(onInstall, 60e3, {
+    reason: 'unbindme'
```
This robotic expansion of literals doesn't make the code more readable. On the contrary, it's less readable in such cases as this one where only one parameter is used.

tophf commented on src/background.js:
```diff
@@ -88,14 +94,18 @@ if ('commands' in chrome) {
   openManage() {
-    openURL({url: '/manage.html'});
+    openURL({
+      url: 'manage.html'
```
This robotic expansion of literals doesn't make the code more readable. On the contrary, it's less readable in such cases as this one where only one parameter is used.

tophf commented on src/background.js:
```diff
@@ -118,24 +128,28 @@ contextMenus = Object.assign({
+      documentUrlPatterns: [URLS.ownOrigin + 'edit*'],
+      click: (info, tab) => {
+        chrome.tabs.sendMessage(tab.id, {
+          method: 'editDeleteText'
```
This robotic expansion of literals doesn't make the code more readable. On the contrary, it's less readable in such cases as this one where only one parameter is used.

tophf commented:
Arrgh! I don't like it. Like really. I'm against the proposed changes to .eslintrc, .editorconfig and the code style.

tophf commented on package.json:
```diff
@@ -0,0 +1,21 @@
+{
```
Now my IDE thinks eslint is not installed. Is it possible to make package.json use my global eslint installation? I don't want to install a local copy of eslint just for stylus.

tophf replied:
Why do we need package.json and package-lock.json at all? My IDE uses the global eslint already.

Mottie replied:
It's odd, sometimes my Atom doesn't catch all the issues... that's why I added eslint. I can remove that bloat, no problem.

tophf commented:
With the exception of edit.* there's no need to change other files, I'm especially against using the automatic conversion because in some cases it's less readable. I've updated eslint locally and I see just a couple of minor spacing/ident issues in all js files except edit.js which produce errors according to the current rules and should be fixed.

Mottie commented:
> manifest.json has been moved

I'll move all the files out of the `src` folder then and I'll spend some time moving the files around as you suggested.

tophf commented:
Wow, eslint v4 changed `indent` rule handling and now it disallows visual aligning of multiline ternaries and some use cases of chained methods previously allowed (issue 8594 in eslint repo) so until they sort it out we'll switch `indent` rule to `indent-legacy`. Please update your .eslintrc here.

Mottie commented:
Ugh...

> Error	ESLint	Definition for rule 'indent-legacy' was not found (indent-legacy)

tophf commented:
Works in eslint 4.1.1 here

Mottie commented:
Ahh ok, I forgot to update it.

Mottie commented:
In the `edit.js` file, this line at 384 can be found:

```js
    const minHeight = cm.defaultTextHeight()
      + cm.display.lineDiv.offsetParent.offsetTop /* .CodeMirror-lines padding */
      + wrapper.offsetHeight - wrapper.clientHeight /* borders */;
```

Eslint is complaining that `'+' should be placed at the end of the line. (operator-linebreak)`; should I fix or disable this issue? Personally, I add my `+` at the end, but I don't know your coding style.

tophf commented:
Yes, `+ at the end is more conventional. You can look at the other js files and see for yourself. I believe there shouldn't be a lot of disagreements so we can discuss each case.

As for edit.js, it doesn't use any coding style, no one refactored/restyled it after we inherited the legacy Stylish code, so apply your best judgment. I don't trust automatic replacement/fixes so all those var-let/const changes should be checked manually to ensure nothing was broken. You know already I dislike expanded literals in case of just one (and relatively short) argument.

Mottie commented:
I ended up starting over
<img title=":_ tuzki_dizzy :" src="https://cloud.githubusercontent.com/assets/136959/13203535/61480e08-d880-11e5-9ea8-7fc03c7d7b56.gif">

Mottie commented:
Oh, one last note... I don't know much about ruby. I'm not sure how or where this require code gets loaded:

```ruby
require_relative '../stylish-chrome-bin/pull_locales_login'
```

Mottie commented:
Updated to fixed conflicts

tophf commented:
* pull_locales* access a file system, not a web page, so the paths should be relative: `../`
  It might be also useful to check both in current dir and `..` so that it's possible to run as `tools/pull_locales` without changing the current dir first.
* messaging.js is a common script, not background-specific, so it should be moved to `js` dir.

Haven't got time yet to check other things.

Mottie commented:
Ok, updated... sorry, I don't know enough python or ruby to be able to make them check both the current and parent directory for a file.

tophf commented on edit/edit.js:
```diff
@@ -0,0 +1,2068 @@
+  }
+  let value = el.type == 'checkbox' ? el.checked : el.value;
+  let url;
+  const themeLink = document.getElementById('cm-theme');
```
I disagree. Just because eslint warns about block scope of let/const doesn't mean it's semantically correct to extract the theme-specific code outside of its case. Either keep `var` or add a eslint override comment

tophf commented on edit/edit.js:
```diff
@@ -0,0 +1,2068 @@
+    tE('heading', 'addStyleTitle');
+    const section = {code: ''};
+    let i;
+    for (i in CssToProperty) {
```
No need to separate the declaration. Also, use `const` because the value isn't changed in its block scope: `for (const i in CssProperty)`

tophf commented on edit/edit.js:
```diff
@@ -0,0 +1,2068 @@
+    const scope = cm ? [cm] : editors;
+    let changed = false;
+    let fixedOldIssues = false;
+    let state;
```
The old variable was local to the next function block, which looks intended despite the same name. Your change makes the subsequent code use a different variable.

Mottie replied:
I agree, maybe renaming the `state` inside of the `scope` function would prevent confusion in the future.

tophf replied:
scopedState?

Mottie replied:
That's what I ended up using 😸 ... oddly, I had to switch the outer `state` definition to use ` var`

```js
var state = cm.state.lint;
```

otherwise there were JS errors inside the `update` function stating it was undefined. Maybe it's too early in the morning, but I can't wrap my head around that one.

tophf replied:
Okay then, we'll handle that in the future: the editor needs a major rewrite anyway.

tophf commented on edit/edit.js:
```diff
@@ -0,0 +1,2068 @@
+    }
+  }
+  function escapeHtml(html) {
+    const chars = {'&': '&amp;', '<': '&lt;', '>': '&gt;', '"': '&quot;', '\'': '&#39;', '/': '&#x2F;'};
```
eslint allows using doublequotes to avoid escaping: `"'"`

tophf commented on edit/edit.js:
```diff
@@ -0,0 +1,2068 @@
+
+function removeAppliesTo(event) {
+  const appliesTo = event.target.parentNode,
+    appliesToList = appliesTo.parentNode;
```
```js
  const appliesTo = event.target.parentNode;
  const appliesToList = appliesTo.parentNode;
```

tophf commented on edit/edit.js:
```diff
@@ -0,0 +1,2068 @@
+function toMozillaFormat() {
+  return getSectionsHashes().map(function(section) {
+    let cssMds = [];
+    let i;
```
No need to separate the declaration. Also, use `const` because the value isn't changed in its block scope: `for (const i in propertyToCss)`

tophf commented:
@schomery, with the above fixed the PR seems okay to me (in the scope defined by its title). Waiting for your review...

Mottie commented:
I added a comment above... too early in the morning for me to figure out why that makes a difference.

tophf commented on edit/edit.js:
```diff
@@ -295,13 +295,13 @@ function acmeEventListener(event) {
     case 'tabSize':
       CodeMirror.setOption('indentUnit', Number(value));
       break;
+    /* eslint-disable no-case-declarations */
```
Hey, what do you think of an alternative "hack", which is probably better than tweaking eslint:

```
case 'theme': {
...........
}
case ......
```

Mottie replied:
Hmm, I've never tried wrapping a case in curly brackets before... if it works, then cool!

tophf replied:
Technically it creates a new block. I think it's better than the non-semantic eslint comments. If you're not opposed then let's use it.

schomery commented:
Is there a reason why we prefer `options.html` outside the `options` directory?

tophf commented:
@schomery, uniformity I guess. All html files are in one place.

schomery commented:
Having the HTML files inside their directories makes more sense to me, but I am okay with the current structure if you guys prefer.

tophf commented:
Moving the main html files into subdirs would break some userstyles that customize pages individually. Do you guys think we can rely on authors to update their styles?

schomery commented:
> Moving the main html files into subdirs would break some userstyles that customize pages individually

we have already renamed some HTML files like index.html -> options.html. This can break styles too.

tophf commented:
Obviously the amount of styles targeting the options page is incomparably lower than those for editor/manage/popup.

schomery commented:
true

narcolepticinsomniac commented:
I can see the concern for userstyle compatibility, but I don't think that should influence organizational decisions. Couldn't the old URL be specified to match whatever you change it to? You could regex out the first letter or two of Stylish's ID for any outliers which use a partial ID in some odd regex.

tophf commented:
@narcolepticinsomniac, do you suggest constantly translating URLs in our background matcher? In that case I don't think adding a persistent internal workaround that will run on *every* URL is better.

narcolepticinsomniac commented:
Ignorant as ever. I figured it'd only be needed on `chrome://` URLs, but obviously not a good idea if it's at all detrimental. Even if there's no viable workaround, you should organize how you see fit IMO. Styles will either adapt, or their compatibility probably isn't great anyway.

schomery commented:
Comparing ESLint with my own JSHint + JSCS. There are a few extra rules I like to use. If they look okay we can add them if ESLint has equivalents

1. http://jscs.info/rule/disallowMultipleLineBreaks
2. http://jshint.com/docs/options/#eqeqeq
3. http://jshint.com/docs/options/#unused
4. http://jshint.com/docs/options/#latedef
5. http://jshint.com/docs/options/#undef

tophf commented:
2. only after you manually check *every* `==` and `!=` in all files to ensure they don't break on conversion to `===`.

These we already use:

1. `no-unexpected-multiline`
3. `no-unused-vars: [1, {args: after-used, vars: local, argsIgnorePattern: ^_}]`
4. `no-use-before-define`
5. `no-undef`

schomery commented:
@tophf can we allow maximum of one space for `space-before-function-paren`?

schomery commented:
Dont you guys prefer:
```
no-multi-spaces: [1]
space-before-function-paren: [1, {
  "anonymous": "always",
  "named": "always",
  "asyncArrow": "always"
}]
```

schomery commented:
> only after you manually check every == and != in all files to ensure they don't break on conversion to ===.

True but why not have a warning; `eqeqeq: [1, always]`

tophf commented:
I agree with `no-multi-spaces: [2, {ignoreEOLComments: true}]` to allow vertically aligned `//` comments

I agree with `space-before-function-paren` for `asyncArrow`.

I definitely dislike `space-before-function-paren` for `named` because there's no point in this rule. Other languages don't add spaces before `(` in these cases so this rule seems really weird.
I can see the point in case of `anonymous` because it's a keyword, but we don't have a lot of `function()` anyway because we use arrow functions normally.

tophf commented:
> True but why not have a warning; eqeqeq: [1, always]

Because it would add a lot of unhelpful noise in IDE. First we need to fix the code, then enable the rule.

schomery commented:
@tophf is there any way to limit spaces before blocks to a single space?
```
function foo()        {

}
```

schomery commented:
any reason for `no-empty-function: [0]`?

tophf commented:
> limit spaces before blocks to a single space

`no-multi-spaces: [2, {ignoreEOLComments: true}]` takes care of that, apparently.

> any reason for no-empty-function: [0]?

It's one of the original rules that I didn't change nor care because there was no offending code. We can change it I guess.

schomery commented:
Okay then. This looks good to me. We can merge if there is no more concerns.

tophf commented:
So the additional rules would be as follows?

```
no-multi-spaces: [2, {ignoreEOLComments: true}]
space-before-function-paren: [2, {anonymous: always, asyncArrow: always, named: never}]
```

The first one is violated in manage.js.
The second one is violated in several files, primarily in edit.js, which would require converting the anonymous functions to arrow syntax (where `this` is not used) or object method shorthand. @Mottie, we're counting on you :-)

As for `eqeqeq`, it'd be nice to handle in this PR (meaning manually checking every occurrence), but I don't think it's critical. There are many high quality js libraries that use `==`.

Mottie commented:
Updated!

Oddly, I added the `no-multi-spaces` rule:

```
no-multi-spaces: [2, {ignoreEOLComments: true}]
```

and I keep getting errors in the Atom dev tools:

> [Linter] Error running ESLint Error: ...\.eslintrc:
	Configuration for rule "no-multi-spaces" is invalid:
	Value "data["0"].ignoreEOLComments" has additional properties.

but  it works just fine when I run `eslint "**/*.js"`... I even updated eslint according to to [issue](https://github.com/airbnb/javascript/issues/1352).

tophf commented:
Seems okay. Let's test it for a while locally before merging. Like for a week or until the next commit, whichever comes first.

tophf commented:
>I keep getting errors in the Atom dev tools:

Simply overwrite the outdated atom's eslint (somewhere in packages\inter-eslint\node_modules) using node's latest one installed via `npm i -g eslint` for example.

narcolepticinsomniac commented:
OK, I'll be using this now for testing.

tophf commented:
@Mottie , if you're still enthusiastic about contributing, you can handle #116 (I'm still mad at AMO) and replace innerHTML assignments with direct DOM manipulation and [construction via $element()](https://github.com/openstyles/stylus/blob/ed83f8f77e00c7ea7f5a8d0d46fb5de8a47cc762/backup/fileSaveLoad.js#L182), see also my comment there about compressing inter-tag spacing and whatnot.

Mottie commented:
I *could* look at it, but I'm still not familiar with the code base and I would get stuck on perplexing code like:

* `cell.innerHTML = cell.textContent;` (in `edit.js`). It's probably easier to completely remove it.
* localization code that is injecting links &ndash; do we modify the locations to remove them?
* CodeMirror's dialog addon uses `innerHTML` to add a template &ndash; do we rewrite that too?

It would be nice if we discussed some of these on a case-by-case basis.

tophf commented on edit.html:
```diff
@@ -1,792 +1,213 @@
+      </li>
+    </template>
+    <template data-id="appliesToEverything">
+      <li class="applies-to-everything" i18n-html="appliesToEverything">
```
`i18n-html` should be `i18n-text` because it's text-only

schomery commented:
I have a suggestion. How about reducing the error level on ESLint to "warning" when the code is not actually breaking things. For instance missing a semicolon should not be displayed as a coding error in the editor.

tophf commented:
A missing semicolon is a serious offence!!111 Worse than innerHTML :-) So I disagree. Any other examples?

schomery commented:
okay then! no more examples :)
